File: src/com/thejoshwa/ultrasonic/androidapp/activity/SelectArtistActivity.java
/*
 This file is part of Subsonic.

 Subsonic is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 Subsonic is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Subsonic.  If not, see <http://www.gnu.org/licenses/>.

 Copyright 2009 (C) Sindre Mehus
 */

package com.thejoshwa.ultrasonic.androidapp.activity;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;

import com.handmark.pulltorefresh.library.PullToRefreshBase;
import com.handmark.pulltorefresh.library.PullToRefreshListView;
import com.handmark.pulltorefresh.library.PullToRefreshBase.OnRefreshListener;
import com.thejoshwa.ultrasonic.androidapp.R;
import com.thejoshwa.ultrasonic.androidapp.domain.Artist;
import com.thejoshwa.ultrasonic.androidapp.domain.Indexes;
import com.thejoshwa.ultrasonic.androidapp.domain.MusicFolder;
import com.thejoshwa.ultrasonic.androidapp.service.MusicService;
import com.thejoshwa.ultrasonic.androidapp.service.MusicServiceFactory;
import com.thejoshwa.ultrasonic.androidapp.util.BackgroundTask;
import com.thejoshwa.ultrasonic.androidapp.util.Constants;
import com.thejoshwa.ultrasonic.androidapp.util.TabActivityBackgroundTask;
import com.thejoshwa.ultrasonic.androidapp.util.Util;
import com.thejoshwa.ultrasonic.androidapp.view.ArtistAdapter;

import java.util.ArrayList;
import java.util.List;

public class SelectArtistActivity extends SubsonicTabActivity implements AdapterView.OnItemClickListener {

    private static final int MENU_GROUP_MUSIC_FOLDER = 10;

    private PullToRefreshListView refreshArtistListView;
    private ListView artistListView;
    private View folderButton;
    private TextView folderName;
    private List<MusicFolder> musicFolders;

    /**
     * Called when the activity is first created.
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.select_artist);

        refreshArtistListView = (PullToRefreshListView) findViewById(R.id.select_artist_list);
        artistListView = refreshArtistListView.getRefreshableView();
        
        refreshArtistListView.setOnRefreshListener(new OnRefreshListener<ListView>() {
            @Override
            public void onRefresh(PullToRefreshBase<ListView> refreshView) {
                new GetDataTask().execute();
            }
        });
        
        artistListView.setOnItemClickListener(this);

        folderButton = LayoutInflater.from(this).inflate(R.layout.select_artist_header, artistListView, false);
        folderName = (TextView) folderButton.findViewById(R.id.select_artist_folder_2);

        if (!Util.isOffline(this)) {
            artistListView.addHeaderView(folderButton);
        }

        registerForContextMenu(artistListView);

        String title = getIntent().getStringExtra(Constants.INTENT_EXTRA_NAME_ALBUM_LIST_TITLE);
        if (title == null) {
        	getActionBar().setSubtitle(Util.isOffline(this) ? R.string.music_library_label_offline : R.string.music_library_label);
        } else {
        	getActionBar().setSubtitle(title);
        }
        
        View browseMenuItem = findViewById(R.id.menu_browse);
        menuDrawer.setActiveView(browseMenuItem);

        musicFolders = null;
        load();
    }
    
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
    	super.onCreateOptionsMenu(menu);
    	return true;
    }

    private void refresh() {
        finish();
        Intent intent = getIntent();
        String title = getIntent().getStringExtra(Constants.INTENT_EXTRA_NAME_ALBUM_LIST_TITLE);
        intent.putExtra(Constants.INTENT_EXTRA_NAME_ALBUM_LIST_TITLE, title);
        intent.putExtra(Constants.INTENT_EXTRA_NAME_REFRESH, true);
        Util.startActivityWithoutTransition(this, intent);
    }

    private void selectFolder() {
        folderButton.showContextMenu();
    }

    private void load() {
        BackgroundTask<Indexes> task = new TabActivityBackgroundTask<Indexes>(this) {
            @Override
            protected Indexes doInBackground() throws Throwable {
                boolean refresh = getIntent().getBooleanExtra(Constants.INTENT_EXTRA_NAME_REFRESH, false);
                MusicService musicService = MusicServiceFactory.getMusicService(SelectArtistActivity.this);
                boolean isOffline = Util.isOffline(SelectArtistActivity.this);
                boolean useId3Tags = Util.getShouldUseId3Tags(SelectArtistActivity.this);
                
                if (!isOffline && !useId3Tags) {
                    musicFolders = musicService.getMusicFolders(refresh, SelectArtistActivity.this, this);
                }
                
                String musicFolderId = Util.getSelectedMusicFolderId(SelectArtistActivity.this);
                
                if (!isOffline && useId3Tags) {
                	return musicService.getArtists(refresh, SelectArtistActivity.this, this);	
                } else {
                	return musicService.getIndexes(musicFolderId, refresh, SelectArtistActivity.this, this);
                }
            }

            @Override
            protected void done(Indexes result) {
            	if (result != null) {
            		List<Artist> artists = new ArrayList<Artist>(result.getShortcuts().size() + result.getArtists().size());
            		artists.addAll(result.getShortcuts());
            		artists.addAll(result.getArtists());
            		artistListView.setAdapter(new ArtistAdapter(SelectArtistActivity.this, artists));
            	}

                // Display selected music folder
                if (musicFolders != null) {
                    String musicFolderId = Util.getSelectedMusicFolderId(SelectArtistActivity.this);
                    if (musicFolderId == null) {
                        folderName.setText(R.string.select_artist_all_folders);
                    } else {
                        for (MusicFolder musicFolder : musicFolders) {
                            if (musicFolder.getId().equals(musicFolderId)) {
                                folderName.setText(musicFolder.getName());
                                break;
                            }
                        }
                    }
                }
            }
        };
        task.execute();
    }

    @Override
    public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
        if (view == folderButton) {
            selectFolder();
        } else {
            Artist artist = (Artist) parent.getItemAtPosition(position);
            Intent intent = new Intent(this, SelectAlbumActivity.class);
            intent.putExtra(Constants.INTENT_EXTRA_NAME_ID, artist.getId());
            intent.putExtra(Constants.INTENT_EXTRA_NAME_NAME, artist.getName());
            Util.startActivityWithoutTransition(this, intent);
        }
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View view, ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu, view, menuInfo);

        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) menuInfo;

        if (artistListView.getItemAtPosition(info.position) instanceof Artist) {
            MenuInflater inflater = getMenuInflater();
            inflater.inflate(R.menu.select_artist_context, menu);
        } else if (info.position == 1) {
            String musicFolderId = Util.getSelectedMusicFolderId(this);
            MenuItem menuItem = menu.add(MENU_GROUP_MUSIC_FOLDER, -1, 0, R.string.select_artist_all_folders);
            if (musicFolderId == null) {
                menuItem.setChecked(true);
            }
            if (musicFolders != null) {
                for (int i = 0; i < musicFolders.size(); i++) {
                    MusicFolder musicFolder = musicFolders.get(i);
                    menuItem = menu.add(MENU_GROUP_MUSIC_FOLDER, i, i + 1, musicFolder.getName());
                    if (musicFolder.getId().equals(musicFolderId)) {
                        menuItem.setChecked(true);
                    }
                }
            }
            
            menu.setGroupCheckable(MENU_GROUP_MUSIC_FOLDER, true, true);
        }
    }

    @Override
    public boolean onContextItemSelected(MenuItem menuItem) {
        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) menuItem.getMenuInfo();

        Artist artist = (Artist) artistListView.getItemAtPosition(info.position);

        if (artist != null) {
            switch (menuItem.getItemId()) {
                case R.id.artist_menu_play_now:
                    downloadRecursively(artist.getId(), false, false, true, false, false, false);
                    break;
                case R.id.artist_menu_play_next:
                    downloadRecursively(artist.getId(), false, false, true, true, false, true);
                    break;
                case R.id.artist_menu_play_last:
                    downloadRecursively(artist.getId(), false, true, false, false, false, false);
                    break;
                case R.id.artist_menu_pin:
                    downloadRecursively(artist.getId(), true, true, false, false, false, false);
                    break;
                default:
                    return super.onContextItemSelected(menuItem);
            }
        } else if (info.position == 1) {
            MusicFolder selectedFolder = menuItem.getItemId() == -1 ? null : musicFolders.get(menuItem.getItemId());
            String musicFolderId = selectedFolder == null ? null : selectedFolder.getId();
            String musicFolderName = selectedFolder == null ? getString(R.string.select_artist_all_folders)
                                                            : selectedFolder.getName();
            Util.setSelectedMusicFolderId(this, musicFolderId);
            folderName.setText(musicFolderName);
            refresh();
        }

        return true;
    }
    
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
			case android.R.id.home:
				menuDrawer.toggleMenu();
				return true; 
            case R.id.main_shuffle:
                Intent intent = new Intent(this, DownloadActivity.class);
                intent.putExtra(Constants.INTENT_EXTRA_NAME_SHUFFLE, true);
                Util.startActivityWithoutTransition(this, intent);
                return true;              
        }

        return false;
    }
    
    private class GetDataTask extends AsyncTask<Void, Void, String[]> {
        @Override
        protected void onPostExecute(String[] result) {
            refreshArtistListView.onRefreshComplete();
            super.onPostExecute(result);
        }

		@Override
		protected String[] doInBackground(Void... params) {
			refresh();
			return null;
		}
    }
}

File: ultrasonic/src/main/kotlin/org/moire/ultrasonic/activity/SelectArtistActivity.kt
/*
 This file is part of Subsonic.

 Subsonic is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 Subsonic is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with Subsonic.  If not, see <http://www.gnu.org/licenses/>.

 Copyright 2020 (C) Jozsef Varga
 */
package org.moire.ultrasonic.activity

import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.PopupMenu
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import org.koin.android.ext.android.inject
import org.koin.android.viewmodel.ext.android.viewModel
import org.moire.ultrasonic.R
import org.moire.ultrasonic.data.ActiveServerProvider
import org.moire.ultrasonic.data.ActiveServerProvider.Companion.isOffline
import org.moire.ultrasonic.domain.Artist
import org.moire.ultrasonic.domain.MusicFolder
import org.moire.ultrasonic.util.Constants
import org.moire.ultrasonic.util.Util

class SelectArtistActivity : SubsonicTabActivity() {
    private val activeServerProvider: ActiveServerProvider by inject()
    private val serverSettingsModel: ServerSettingsModel by viewModel()
    private val artistListModel: ArtistListModel by viewModel()

    private var refreshArtistListView: SwipeRefreshLayout? = null
    private var artistListView: RecyclerView? = null
    private var musicFolders: List<MusicFolder>? = null
    private lateinit var viewManager: RecyclerView.LayoutManager
    private lateinit var viewAdapter: ArtistRowAdapter

    /**
     * Called when the activity is first created.
     */
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.select_artist)

        refreshArtistListView = findViewById(R.id.select_artist_refresh)
        refreshArtistListView!!.setOnRefreshListener {
            artistListModel.refresh(refreshArtistListView!!)
        }

        val shouldShowHeader = (!isOffline(this) && !Util.getShouldUseId3Tags(this))

        val title = intent.getStringExtra(Constants.INTENT_EXTRA_NAME_ALBUM_LIST_TITLE)
        if (title == null) {
            setActionBarSubtitle(
                if (isOffline(this)) R.string.music_library_label_offline
                else R.string.music_library_label
            )
        } else {
            actionBarSubtitle = title
        }
        val browseMenuItem = findViewById<View>(R.id.menu_browse)
        menuDrawer.setActiveView(browseMenuItem)
        musicFolders = null

        val refresh = intent.getBooleanExtra(Constants.INTENT_EXTRA_NAME_REFRESH, false)

        artistListModel.getMusicFolders(refresh, refreshArtistListView!!)
            .observe(
                this,
                Observer { changedFolders ->
                    if (changedFolders != null) {
                        musicFolders = changedFolders
                        viewAdapter.setFolderName(getMusicFolderName(changedFolders))
                    }
                }
            )

        val artists = artistListModel.getArtists(refresh, refreshArtistListView!!)
        artists.observe(
            this, Observer { changedArtists -> viewAdapter.setData(changedArtists) }
        )

        viewManager = LinearLayoutManager(this)
        viewAdapter = ArtistRowAdapter(
            artists.value ?: listOf(),
            getText(R.string.select_artist_all_folders).toString(),
            shouldShowHeader,
            { artist -> onItemClick(artist) },
            { menuItem, artist -> onArtistMenuItemSelected(menuItem, artist) },
            { view -> onFolderClick(view) }
        )

        artistListView = findViewById<RecyclerView>(R.id.select_artist_list).apply {
            setHasFixedSize(true)
            layoutManager = viewManager
            adapter = viewAdapter
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                menuDrawer.toggleMenu()
                return true
            }
            R.id.main_shuffle -> {
                val intent = Intent(this, DownloadActivity::class.java)
                intent.putExtra(Constants.INTENT_EXTRA_NAME_SHUFFLE, true)
                startActivityForResultWithoutTransition(this, intent)
                return true
            }
        }
        return false
    }

    private fun getMusicFolderName(musicFolders: List<MusicFolder>): String {
        val musicFolderId = activeServerProvider.getActiveServer().musicFolderId
        if (musicFolderId != null && musicFolderId != "") {
            for ((id, name) in musicFolders) {
                if (id == musicFolderId) {
                    return name
                }
            }
        }
        return getText(R.string.select_artist_all_folders).toString()
    }

    private fun onItemClick(artist: Artist) {
        val intent = Intent(this, SelectAlbumActivity::class.java)
        intent.putExtra(Constants.INTENT_EXTRA_NAME_ID, artist.id)
        intent.putExtra(Constants.INTENT_EXTRA_NAME_NAME, artist.name)
        intent.putExtra(Constants.INTENT_EXTRA_NAME_PARENT_ID, artist.id)
        intent.putExtra(Constants.INTENT_EXTRA_NAME_ARTIST, true)
        startActivityForResultWithoutTransition(this, intent)
    }

    private fun onFolderClick(view: View) {
        val popup = PopupMenu(this, view)

        val musicFolderId = activeServerProvider.getActiveServer().musicFolderId
        var menuItem = popup.menu.add(
            MENU_GROUP_MUSIC_FOLDER, -1, 0, R.string.select_artist_all_folders
        )
        if (musicFolderId == null || musicFolderId.isEmpty()) {
            menuItem.isChecked = true
        }
        if (musicFolders != null) {
            for (i in musicFolders!!.indices) {
                val (id, name) = musicFolders!![i]
                menuItem = popup.menu.add(MENU_GROUP_MUSIC_FOLDER, i, i + 1, name)
                if (id == musicFolderId) {
                    menuItem.isChecked = true
                }
            }
        }
        popup.menu.setGroupCheckable(MENU_GROUP_MUSIC_FOLDER, true, true)

        popup.setOnMenuItemClickListener { item -> onFolderMenuItemSelected(item) }
        popup.show()
    }

    private fun onArtistMenuItemSelected(menuItem: MenuItem, artist: Artist): Boolean {
        when (menuItem.itemId) {
            R.id.artist_menu_play_now ->
                downloadRecursively(artist.id, false, false, true, false, false, false, false, true)
            R.id.artist_menu_play_next ->
                downloadRecursively(artist.id, false, false, true, true, false, true, false, true)
            R.id.artist_menu_play_last ->
                downloadRecursively(artist.id, false, true, false, false, false, false, false, true)
            R.id.artist_menu_pin ->
                downloadRecursively(artist.id, true, true, false, false, false, false, false, true)
            R.id.artist_menu_unpin ->
                downloadRecursively(artist.id, false, false, false, false, false, false, true, true)
            R.id.artist_menu_download ->
                downloadRecursively(artist.id, false, false, false, false, true, false, false, true)
        }
        return true
    }

    private fun onFolderMenuItemSelected(menuItem: MenuItem): Boolean {
        val selectedFolder = if (menuItem.itemId == -1) null else musicFolders!![menuItem.itemId]
        val musicFolderId = selectedFolder?.id
        val musicFolderName = selectedFolder?.name
            ?: getString(R.string.select_artist_all_folders)
        if (!isOffline(this)) {
            val currentSetting = activeServerProvider.getActiveServer()
            currentSetting.musicFolderId = musicFolderId
            serverSettingsModel.updateItem(currentSetting)
        }
        viewAdapter.setFolderName(musicFolderName)
        artistListModel.refresh(refreshArtistListView!!)
        return true
    }

    companion object {
        private const val MENU_GROUP_MUSIC_FOLDER = 10
    }
}
